### PR TITLE
Add an initial QUIC Technical requirements document

### PR DESCRIPTION
**Pull request by mattcaswell**
We started to write down some technical requirements in various places. This just brings them all into one place. These are not yet agreed by anyone and are just a starting point. I expect the list to grow over time.

### CODE REVIEW

**Comment by mattcaswell** [comment]
**Question for OTC: Does the OTC agree with these requirements?**

**Comment by kaduk** [comment]
I'm not sure how much duplication of comments made elsewhere on similar content you're looking for/not looking for.
But I'll reiterate that I continue to think this list of requirements misses the mark in how it handles single-stream vs multi-stream.  The QUIC stream "namespace" is huge, with more streams available than a connection could reasonably use even if it stays up for years.  The current major protocols that are being written over/ported to use QUIC make use of that flexibility to use the stream as a demarcator, for example with a singl stream corresponding to an HTTP or DNS request/response pair and then being closed.  This gives the application clear protection from truncation and request-smuggling attacks and can simplify the necessary framing layer for the messages being conveyed.
I don't see what evidence is being used to support the stance that there is a large body of applications/protocols out there that are going to want to switch from single-stream TLS 1.3 to single-stream QUIC in the way that's being presumed here.  If you control both endpoints enough to be able to do that, you control them enough to be able to make bigger changes and reap the full benefits of QUIC, and if you don't control the endpoints that strongly there's not a very clear path to using QUIC at all, so the minimal API differences proposed as a requirement here would go unused.
I think it would be incredibly useful to hear more about what data is being used to come to the conclusion that there is actual substantial demand for the sort of "minimal porting procedures" for an application to move to single-stream QUIC as presented here.  With the data I have from the IETF, it just looks like it's going to be wasted effort on our part.

**Comment by tmshort** [comment]
For those who are unaware, @kaduk is completing his second stint as one of the IETF Security Area Directors. His review is required for practically *every single* security-related IETF internet-draft before it becomes an RFC. He is well-versed in QUIC, who is implementing it, and what they are doing with it.

**Comment by mattcaswell** [comment]
I've added a summary of the original OMC requirements, as well as a set of requirements extracted from the OMC blog post. I've also pulled out a section at the end to summarise the requirements that are specific to the MVP.

**Comment by paulidale** [comment]
Action @openssl/otc read the document!

**Comment by mattcaswell** [comment]
Updated following the OTC review of this last week. Please take another look. I aim to start a vote on this in a couple of days.

File: doc/designs/quic-design/quic-requirements.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+* The platform policy, covering the primary and secondary platforms, should be
+  followed. (Note that this includes testing of primary and secondary platforms
+  on project CI)
+
+OMC Blog post requirements
+--------------------------
+
+The OMC additionally published a
+[blog post](https://www.openssl.org/blog/blog/2021/11/25/openssl-update/) which
+also contained somerequirements regarding QUIC. Statements from that blog post
```
Comment (kroeckx) [inline doc/designs/quic-design/quic-requirements.md:83]: missing space

**Comment by mattcaswell** [comment]
OTC voted to accept this. See openssl/technical-policies#37. Removing the "need OTC decision" label.

**Review by t-j-h — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Review by ccbrantley — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks!

**Pull request closed without merging**